Formalize Provider config and doc-gen.

jsirois opened the pull request:
This paves the way for the lift manifest format to be doc-gen'd as well.

jsirois commented:
There was no functionality change, but the delta is large; so I also ran this over in scie-pants with success:
```
$ cargo run -p package -- --science ../../a-scie/lift test --tools-pex-mismatch-warn
```

The new doc looks like so:
![Screenshot 2023-06-03 160121](https://github.com/a-scie/lift/assets/157586/537d1032-df39-45d4-ab36-4b495e532e58)

jsirois commented on science/data.py:
```diff
@@ -0,0 +1,120 @@
+
+
+@dataclass(frozen=True)
+class Data:
```
This was extracted with almost no change from `science/config.py` above.

jsirois commented on science/model.py:
```diff
@@ -247,24 +259,89 @@ def expand_placeholder(match: Match) -> str:
+                f"{subject} cannot use the reserved binding names: {', '.join(reserved_conflicts)}"
+            )
+
+    def __post_init__(self) -> None:
```
These `__post_init__` validations were moved ~unchanged from `science/config.py`.

sureshjoshi commented on science/providers/python_build_standalone.py:
```diff
@@ -49,18 +52,61 @@ def with_fingerprint(self, fingerprint: Fingerprint) -> FingerprintedAsset:
+    )
+    release: str | None = dataclasses.field(
+        default=None,
+        metadata=help_doc(
```
Think it's worth also mentioning an example release and citing the same URL as above?
https://github.com/indygreg/python-build-standalone/releases

e.g `20230507`

I got burned by this my first time around.

jsirois replied:
Definitely. Done.

I re-worked the field sections so that they are linkable. The screen shot is updated - note the link tree in the contents siderbar summarizing this new heading structure.

jsirois commented:
@sureshjoshi I think I've addressed your feedback. It should be easier to not go astray now if you happen to read these docs. I'm going to push ahead and merge. I'd like to finish this doc detour today and would like to get lift manifest format doc-gen and `science doc open` for local docs correct-for-your-version both in the 0.2.0 release.

jsirois commented:
https://science.scie.app/providers.html
